runtime/src/accounts_db.rs

```diff
@@ -1801,6 +2309,96 @@ impl AccountsDB {
+    fn purge_removed_slots_from_store(&self, removed_slots: &HashSet<Slot>) {
+        // Check all slots `removed_slots` are no longer rooted
+        let mut safety_checks_elapsed = Measure::start("safety_checks_elapsed");
+        for slot in removed_slots.iter() {
+            assert!(!self.accounts_index.is_root(*slot))
+        }
+        safety_checks_elapsed.stop();
+
+        // Purge the storage entries of the removed slots
+        let mut remove_storages_elapsed = Measure::start("remove_storages_elapsed");
+        let mut all_removed_slot_storages = vec![];
+        let mut total_removed_storage_entries = 0;
+        let mut total_removed_bytes = 0;
+        for slot in removed_slots {
+            // The removed slot must alrady have been flushed from the cache
+            assert!(self.accounts_cache.slot_cache(*slot).is_none());
+            if let Some((_, slot_removed_storages)) = self.storage.0.remove(&slot) {
+                {
+                    let r_slot_removed_storages = slot_removed_storages.read().unwrap();
+                    total_removed_storage_entries += r_slot_removed_storages.len();
+                    total_removed_bytes += r_slot_removed_storages
+                        .values()
+                        .map(|i| i.accounts.capacity())
+                        .sum::<u64>();
+                }
+                all_removed_slot_storages.push(slot_removed_storages.clone());
+            }
+        }
+        remove_storages_elapsed.stop();
+
+        let num_slots_removed = all_removed_slot_storages.len();
+
+        let recycle_stores_write_time =
+            self.recycle_slot_stores(total_removed_storage_entries, &all_removed_slot_storages);
+
+        let mut drop_storage_entries_elapsed = Measure::start("drop_storage_entries_elapsed");
+        // Backing mmaps for removed storages entries explicitly dropped here outside
+        // of any locks
+        drop(all_removed_slot_storages);
+        drop_storage_entries_elapsed.stop();
+
+        datapoint_info!(
+            "purge_slots_time",
+            ("safety_checks_elapsed", safety_checks_elapsed.as_us(), i64),
+            (
+                "remove_storages_elapsed",
+                remove_storages_elapsed.as_us(),
+                i64
+            ),
+            (
+                "drop_storage_entries_elapsed",
+                drop_storage_entries_elapsed.as_us(),
+                i64
+            ),
+            ("num_slots_removed", num_slots_removed, i64),
+            (
+                "total_removed_storage_entries",
+                total_removed_storage_entries,
+                i64
+            ),
+            ("total_removed_bytes", total_removed_bytes, i64),
+            (
+                "recycle_stores_write_elapsed",
+                recycle_stores_write_time,
+                i64
+            ),
+        );
+    }
+
+    fn purge_slot_cache_keys(&self, slot: Slot, slot_cache: SlotCache) -> Vec<(u64, AccountInfo)> {
```

Comment:
nits: `dead_slot: Slot`.

Reply:
also, nits: `fn purge_dead_slot_cache`,

so, this will be like:

`fn purge_dead_slot_cache(&self, dead_slot: Slot, slot_cache: SlotCache) -> Vec<(u64, AccountInfo)> `